### PR TITLE
TokenEffect getStackDescription enhancement (for Indulge // Excess)

### PR DESCRIPTION
**Pull request by Northmoc**
The other work is now covered in #622

### CODE REVIEW

File: forge-game/src/main/java/forge/game/ability/AbilityUtils.java [hunk 1]
```diff
@@ -2103,6 +2103,15 @@ public static int xCount(final Card c, final String s, final CardTraitBase ctb)
             }
             return doXMath(sum, expr, c, ctb);
         }
+
+        if (sq[0].startsWith("NumDealtCombatDamageToPlayersThisTurn")) {
+            int amt = 0;
+            for (Player p : game.getPlayersAttackedThisTurn().get(player)) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/AbilityUtils.java:2109]: What if getPlayersAttackedThisTurn returns null?

**Comment by tool4ever** [comment]
The problem is these damageMaps are only keyed with card ID.

That means if you get two combats and attack with the same creature in both, it will only be counted once _if you flicker the creature between them_.

Not sure yet what size of approach is best to fix that. It will probably need to be converted into LKI pairs again like I did for some other stuff - and then only those entries with `equalsWithTimestamp` must be merged.

@Hanmac
Maybe we should have some data class to handle these cases?

**Comment by Hanmac** [comment]
ugh i will look into some data classes

**Comment by Hanmac** [comment]
> That class looks interesting, nice find
> 
> `getAssignedDamageSources()` is still used by _Inferno Trap_
> 
> if we switch to LKI we could rewrite `getAssignedDamage(final String type)` to use `assignedCombatDamage` and use that for Prowl instead of `addProwlType`
> 
> it would also help in fixing _Admiral Beckett Brass_

stupid github, it chouldn't find YourDamageSourcesThisTurn while i was on your Fork

as for LKI, wouldn't help us with `Equivalence.Wrapper` because the same creature could have changed its types a second time it deals combat damage, so the prowl check would fail, so we can't use Map there.

**Review by tool4ever — APPROVED** [review]

**Pull request merged**